### ISSUE TITLE
Don't mark messages as permanently failed unless they are irredeemable

### ISSUE THREAD

**Issue by SCdF** [medic/medic-gateway#139]
Currently gateway marks a message as failed if it fails to send, and if a message has failed it is not retried. Many of these failures can be temporary however, including things like being out of credit, or there being a generic network problem.

We should change these semantics so that we retry failures that we consider redeemable. The exact parameters around this are up for discussion, but something like this:
 - When gateway detects a failure it marks the message as failed (this is as it is now)
 - When API is informed that a message has failed it marks it as such, and also increments a failure count by 1 (or to 1).
 - When gateway requests messages to be sent out, api returns all messages that it currently would, but also messages that failed over 6 hours ago, that have not failed more than 12 times (12 * 6 = 72hrs = 3 days = enough time to top up your phone).

Things we'd have to check:
 - That gateway will actually re-send messages that we supply it that have failed before (ie maybe there are code changes in gateway that we'd have to make too)
 - Which failures are irredeemable and which are not. AFAICT maybe none are, especially if you can change a phone number of a message once it's failed (e.g. you try to send a message to a contact out and it fails because the number is invalid, so you change the contact and 6 hours later it sends to the correct number)
 - If this will incur any more charges. e.g, are all of these failures "free", or are some of them going to cost the price of an SMS to send.

This ticket is important, like #4715, because it contributes to a need to over-manage gateway phones. Specifically in this case, because running out of credit is considered a hard failure, TLs / PMs need to micromanage how many credit is on the phones to avoid having legitimate messages fail to send.

**Comment by garethbowen** [comment]
Not yet started - delaying to 3.9.0

**Comment by garethbowen** [comment]
Moved to 3.11.0 to align with other gateway work.

**Comment by latin-panda** [comment]
I couldn't add any [new error code](https://developer.android.com/reference/android/telephony/SmsManager#RESULT_RIL_NETWORK_ERR) for SMS, since we are supporting older versions of Android and most of the nice ones were added in SDK 30.

**Comment by dianabarsan** [comment]
This is ready for AT on `139-dont-mark-sms-permanently-failed`. 
This branch is based on the branch that work for this issue was done on: https://github.com/medic/medic-gateway/issues/154
So please review and get  #154 merged first before AT-ing this.

**Comment by latin-panda** [comment]
### Summary
When an error happens after sending a SMS, there are 2 possible actions: to soft-fail or to hard-fail.
When an error hard-fail, Gateway won't send again the SMS automatically, the user needs to manually select it and click on 'resend'.
When an error soft-fail, Gateway will try to resend again the SMS. We are keeping count of the times Gateway is resending in WOM_clmRETRIES of tblWO_MESSAGE table, this count is being printed in the app's logs.

From the [documentation](https://github.com/medic/medic-gateway/pull/159/files#diff-04c6e90faac2675aa89e2176d2eec7d8):

#### SMS Retry Mechanism

Gateway will retry to send the SMS when any of these errors occurs: `RESULT_ERROR_NO_SERVICE`, `RESULT_ERROR_NULL_PDU` and `RESULT_ERROR_RADIO_OFF`.

1. A possible temporary error occurs and Gateway retries sending the SMS:
    1.1 SMS status will be updated to `UNSENT`, so Gateway will find it and add it into the `send queue` automatically.
    1.2 SMS' `retry counter` increases by 1.
    1.3 The retry attempt is scheduled based on this formula: `SMS' last activity time + ( 1 minute * (retry counter ^ 1.5) )`. This means the time between retries is incremental.
    1.4 Gateway logs: the error, the retry counter and the retry scheduled time. Sample: `Sending SMS to +1123123123 failed (cause: radio off) Retry #5 in 15 min`

2. Gateway has a maximum limit of attempts to retry sending SMS (currently 20), If this is reached then:
    2.1 Gateway will hard fail the SMS by updating its status to `FAILED` and won't retry again.
    2.2 Gateway logs error. Sample: `Sending message to +1123123123 failed (cause: radio off) Not retrying`

3. At this point the user has the option of manually select the SMS and press `Retry` button.
    3.1 If they do and SMS fails again, then the process will restart from step # 1.

**Comment by newtewt** [comment]
@latin-panda  we were able to test this what we feel is not a full test but does cover the functionalities. Full testing would be hitting all the failure states but we were not able to get ways to force them all. 

For hard failures, we used a phone that had no sim card and it caused a generic error right away. Forcing it into the hard failure scenario and had to select the message for retry and send again. 

For Soft Failure, we used a phone that was connected to wifi but airplane mode otherwise. This causes the `RESULT_ERROR_RADIO_OFF` error. After the retry amount was hit this was then put into a hard failed state. Selecting the message and retrying attempted the resend. 

A second phone we were able to get the soft failure for `RESULT_ERROR_NO_SERVICE` by removing the SIM card.  Then eventually into the hard fail after the try limit.  After the hard fail the SIM was put back into the phone. Then we retried and it sent the message. We were noticing the message state being shown as unsent until refreshing. We will open a new ticket for this issue. 

All of the details were displayed in the event  log as well.
